crates/iceberg/src/spec/values.rs

```diff
@@ -996,10 +1008,552 @@ mod timestamptz {
     }
 }
 
+mod serde {
+    use std::collections::BTreeMap;
+
+    use crate::{
+        spec::{PrimitiveType, Type},
+        Error, ErrorKind,
+    };
+
+    use super::{Literal, PrimitiveLiteral};
+    use serde::{
+        de::Visitor,
+        ser::{SerializeMap, SerializeSeq, SerializeStruct},
+        Deserialize, Serialize,
+    };
+    use serde_bytes::ByteBuf;
+    use serde_derive::Deserialize as DeserializeDerive;
+    use serde_derive::Serialize as SerializeDerive;
+
+    #[derive(SerializeDerive, DeserializeDerive)]
+    #[serde(transparent)]
+    /// Raw literal representation used for serde. The serialize way is used for Avro serializer.
+    pub struct RawLiteral(RawLiteralEnum);
+
+    impl RawLiteral {
+        /// Covert literal to raw literal.
+        pub fn try_from(literal: Literal, ty: &Type) -> Result<Self, Error> {
+            Ok(Self(RawLiteralEnum::try_from(literal, ty)?))
+        }
+
+        /// Convert raw literal to literal.
+        pub fn try_into(self, ty: &Type) -> Result<Option<Literal>, Error> {
+            self.0.try_into(ty)
+        }
+    }
+
+    #[derive(SerializeDerive, Clone)]
+    #[serde(untagged)]
+    enum RawLiteralEnum {
+        Null,
+        Boolean(bool),
+        Int(i32),
+        Long(i64),
+        Float(f32),
+        Double(f64),
+        String(String),
+        Bytes(ByteBuf),
+        List(List),
+        StringMap(StringMap),
+        Record(Record),
+    }
+
+    #[derive(Clone)]
+    struct Record {
+        required: Vec<(String, RawLiteralEnum)>,
+        optional: Vec<(String, Option<RawLiteralEnum>)>,
```

Comment:
Why we need to distinguish `required` or `optional` here? I think we should check them when converting from `Literal` to `RawLiteralEnum`? And in `Record` we only need to simply keep `Vec<(String, Option<RawLiteralEnum)>`?

Reply:
Because we need identify required when serialize `RawLiteral`. If we simply keep `Vec<(String, Option<RawLiteralEnum)>`, we need somthing like `Vec<(String, (bool,Option<RawLiteralEnum))>` to identify whether it's required. Seems it will cost more space than keep required and optional.🤔

Reply:
I still don't get the point here. Is there a distinction for serializer when they implementing `serialize_i32(Some(1))` or `serialize_i32(1)`?

Reply:
Yes. 
- For `serialize_i32(Some(1))`, we get the avro value like `Union(1,null)`
- For `serialize_i32(1)`, we get the avro value like `1`
They correspond different schema.

Reply:
I see, thanks for explaination.